File: src/main/java/no/novari/configuration/validation/ConfigurationValidatorFactory.java
package no.novari.configuration.validation;

import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import no.novari.configuration.kafka.InstanceMetadataRequestProducerService;
import no.novari.configuration.kafka.IntegrationRequestProducerService;
import no.novari.configuration.kafka.MetadataRequestProducerService;
import no.novari.configuration.model.integration.Integration;
import no.novari.configuration.model.metadata.InstanceMetadataCategory;
import no.novari.configuration.model.metadata.InstanceMetadataContent;
import no.novari.configuration.model.metadata.InstanceValueMetadata;
import no.novari.configuration.model.metadata.IntegrationMetadata;
import org.hibernate.validator.HibernateValidatorFactory;
import org.springframework.stereotype.Service;

import java.util.AbstractMap;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class ConfigurationValidatorFactory {

    private final ValidatorFactory validatorFactory;

    private final IntegrationRequestProducerService integrationRequestProducerService;
    private final MetadataRequestProducerService metadataRequestProducerService;
    private final InstanceMetadataRequestProducerService instanceMetadataRequestProducerService;

    public ConfigurationValidatorFactory(
            ValidatorFactory validatorFactory,
            IntegrationRequestProducerService integrationRequestProducerService,
            MetadataRequestProducerService metadataRequestProducerService,
            InstanceMetadataRequestProducerService instanceMetadataRequestProducerService
    ) {
        this.validatorFactory = validatorFactory;
        this.integrationRequestProducerService = integrationRequestProducerService;
        this.metadataRequestProducerService = metadataRequestProducerService;
        this.instanceMetadataRequestProducerService = instanceMetadataRequestProducerService;
    }

    public Validator getValidator(Long integrationId, Long metadataId) {
        return validatorFactory
                .unwrap(HibernateValidatorFactory.class)
                .usingContext()
                .constraintValidatorPayload(getConfigurationValidationContext(integrationId, metadataId))
                .getValidator();
    }

    private ConfigurationValidationContext getConfigurationValidationContext(Long integrationId, Long metadataId) {

        Integration integration = integrationRequestProducerService
                .get(integrationId)
                .orElseThrow(() -> new CouldNotFindIntegrationException(integrationId));

        IntegrationMetadata metadata = metadataRequestProducerService
                .get(metadataId)
                .orElseThrow(() -> new CouldNotFindMetadataException(metadataId));

        InstanceMetadataContent instanceMetadata = instanceMetadataRequestProducerService
                .get(metadataId)
                .orElseThrow(() -> new CouldNotFindInstanceMetadataException(metadataId));

        return ConfigurationValidationContext
                .builder()
                .integration(integration)
                .metadata(metadata)
                .instanceValueTypePerKey(getInstanceValueTypePerKey(instanceMetadata))
                .build();

    }

    private Map<String, InstanceValueMetadata.Type> getInstanceValueTypePerKey(
            InstanceMetadataContent instanceMetadata
    ) {
        return getInstanceValueTypePerKeyEntries(instanceMetadata)
                .stream()
                .collect(Collectors.toMap(
                        AbstractMap.SimpleEntry::getKey,
                        AbstractMap.SimpleEntry::getValue
                ));
    }

    private List<AbstractMap.SimpleEntry<String, InstanceValueMetadata.Type>> getInstanceValueTypePerKeyEntries(
            InstanceMetadataContent instanceMetadataContent
    ) {
        return Stream.concat(
                        instanceMetadataContent
                                .getCategories()
                                .stream()
                                .map(InstanceMetadataCategory::getContent)
                                .map(this::getInstanceValueTypePerKeyEntries)
                                .flatMap(Collection::stream),
                        instanceMetadataContent
                                .getInstanceValueMetadata()
                                .stream()
                                .map(instanceValueMetadata -> new AbstractMap.SimpleEntry<>(
                                        instanceValueMetadata.getKey(),
                                        instanceValueMetadata.getType()
                                ))
                )
                .toList();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/ConfigurationValidatorFactory.kt
package no.novari.flyt.configuration.validation

import jakarta.validation.Validator
import jakarta.validation.ValidatorFactory
import no.novari.flyt.configuration.kafka.InstanceMetadataRequestProducerService
import no.novari.flyt.configuration.kafka.IntegrationRequestProducerService
import no.novari.flyt.configuration.kafka.MetadataRequestProducerService
import no.novari.flyt.configuration.model.integration.Integration
import no.novari.flyt.configuration.model.metadata.InstanceMetadataContent
import no.novari.flyt.configuration.model.metadata.InstanceValueMetadata
import no.novari.flyt.configuration.model.metadata.IntegrationMetadata
import org.hibernate.validator.HibernateValidatorFactory
import org.springframework.stereotype.Service
import java.util.AbstractMap

@Service
class ConfigurationValidatorFactory(
    private val validatorFactory: ValidatorFactory,
    private val integrationRequestProducerService: IntegrationRequestProducerService,
    private val metadataRequestProducerService: MetadataRequestProducerService,
    private val instanceMetadataRequestProducerService: InstanceMetadataRequestProducerService,
) {
    fun getValidator(
        integrationId: Long,
        metadataId: Long,
    ): Validator =
        validatorFactory
            .unwrap(HibernateValidatorFactory::class.java)
            .usingContext()
            .constraintValidatorPayload(getConfigurationValidationContext(integrationId, metadataId))
            .getValidator()

    private fun getConfigurationValidationContext(
        integrationId: Long,
        metadataId: Long,
    ): ConfigurationValidationContext {
        val integration: Integration =
            integrationRequestProducerService.get(integrationId)
                ?: throw CouldNotFindIntegrationException(integrationId)

        val metadata: IntegrationMetadata =
            metadataRequestProducerService.get(metadataId)
                ?: throw CouldNotFindMetadataException(metadataId)

        val instanceMetadata: InstanceMetadataContent =
            instanceMetadataRequestProducerService.get(metadataId)
                ?: throw CouldNotFindInstanceMetadataException(metadataId)

        return ConfigurationValidationContext
            .builder()
            .integration(integration)
            .metadata(metadata)
            .instanceValueTypePerKey(getInstanceValueTypePerKey(instanceMetadata))
            .build()
    }

    private fun getInstanceValueTypePerKey(
        instanceMetadata: InstanceMetadataContent,
    ): Map<String, InstanceValueMetadata.Type> =
        getInstanceValueTypePerKeyEntries(instanceMetadata)
            .associate { entry -> entry.key to entry.value }

    private fun getInstanceValueTypePerKeyEntries(
        instanceMetadataContent: InstanceMetadataContent,
    ): List<AbstractMap.SimpleEntry<String, InstanceValueMetadata.Type>> {
        val categoryEntries =
            instanceMetadataContent.categories
                .mapNotNull { it.content }
                .flatMap { getInstanceValueTypePerKeyEntries(it) }

        val valueEntries =
            instanceMetadataContent.instanceValueMetadata
                .mapNotNull { instanceValueMetadata ->
                    val key = instanceValueMetadata.key
                    val type = instanceValueMetadata.type
                    if (key == null || type == null) null else AbstractMap.SimpleEntry(key, type)
                }

        return categoryEntries + valueEntries
    }
}
